documentation/general/dash-sai-pipeline-packet-flow.md

```diff
@@ -0,0 +1,886 @@
+    CTL0 --> |Flow Hit| AA0
+    PreACL0 --> MATS0
+    MATS0 --> AA0
+    AA0 --> |Flow Miss| PostACL0
+    PostACL0 --> CTU0
+    AA0 --> |Flow Hit| MS0
+    AA0 --> |Flow Change| CTU0
+    CTU0 --> MS0
+    MS0 --> Out
+
+    CTL1 --> |Flow Miss| PreACL1
+    CTL1 --> |Flow Hit| AA1
+    PreACL1 --> MATS1
+    MATS1 --> AA1
+    AA1 --> |Flow Miss| PostACL1
+    PostACL1 --> CTU1
+    AA1 --> |Flow Hit| MS1
+    AA1 --> |Flow Change| CTU1
+    CTU1 --> MS1
+    MS1 --> Out
+```
+
+This design allows our upper layers to be flexible and doesn't limit to any specific object models for modeling their own policies. For example, [SONiC-DASH pipeline](https://github.com/sonic-net/SONiC/blob/master/doc/dash/dash-sonic-hld.md#2-packet-flows) uses DASH-SAI pipeline to implement VNET routing scenarios by translating the [SONiC DASH APIs](https://github.com/sonic-net/sonic-dash-api/tree/master/proto) to DASH-SAI model:
+
+- A DASH-SAI pipeline is used to represent a VM NIC (ENI).
+- The VxLAN VNI is used to do the direction lookup.
+- The inner MAC address is used for pipeline lookup, a.k.a. ENI lookup (ENI).
+- Once it enters the corresponding DASH pipeline, the `outbound` pipeline will be used to process the packets sourced from the VM, while the `inbound` pipeline will be used to process the packets destined to the VM.
+
+## 5. Pipeline components
+
+### 5.1. Per-packet metadata bus
+
+First of all, since we have multiple matching stages in the pipeline, we need a way to pass the information from the matched entries in the earlier stages to the later ones to help us making final decisions on packet transformation. And this is what metadata bus is for.
+
+At high-level, the metadata bus is a set of fields that being carried all the way through the pipeline along with the packet. It contains:
+
+- The information from the original packet, such as encap information.
+- The information from each matched entry and related data structures, e.g., when a VNET mapping entry is matched, we will publish the information from VNET to the metadata bus.
+- The routing action and its inline parameters from the matched entry.
+
+Implementation-wise, this is similar to Packet Header Vector or Bus in NPL.
+
+### 5.2. Packet parsing
+
+In ASIC, parser is usually the first stage of the pipeline. It parses the incoming packet and extract the information into metadata bus, such as the source and destination MAC, IP, etc. And based on these information, we will invoke the rest of pipeline.
+
+#### 5.2.1. Multi-layer encap parsing
+
+Multi-layer encap brings the problem in the packet parsing, because of encap layer detection and overlay packet handling:
+
+1. In ASIC, parser is always parsing from the outermost bits to innermost bits. However, in DASH, the encap stack is defined reversely: underlay1 -> underlay0 -> overlay. Because the incoming packet could have various number of encaps, we need to be able to map the encaps to the right layer.
```

Comment:
```suggestion
1. In an ASIC, the parser is always parsing from the outermost bits to innermost bits. However, in DASH, the encap stack is defined reversely: underlay1 -> underlay0 -> overlay. Because the incoming packet could have various number of encaps, we need to be able to map the encaps to the right layer.
```